File: 2020/puzzle3.hs
import System.IO

main :: IO ()
main = do
  lines <- readLinesFromFile "puzzle3.dat"
  let slopes = [(1,1), (3,1), (5,1), (7,1), (1,2)]
  putStrLn $ "Part 1: " ++ show (countTrees lines $ slopes !! 1)
  putStrLn $ "Part 2: " ++ show (product $ map (countTrees lines) slopes)

readLinesFromFile :: FilePath -> IO [String]
readLinesFromFile filePath = do
  contents <- readFile filePath
  return (lines contents)

countTrees :: [String] -> (Int, Int) -> Int
countTrees lines (dx, dy) = do
  let ylist = [0,dy..((length lines) - 1)]
  let xlist = map (`mod` length (head lines)) [0,dx..]
  let coords = zip xlist ylist
  let trees = filter (\(x, y) -> (lines !! y) !! x == '#') coords
  length trees

File: 2020/puzzle3.py

slopes = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]  # list of slopes to check

# function to read "puzzle3.dat" as a list of string, stripping spaces and newlines
def read_file():
    with open("puzzle3.dat", "r") as f:
        return [line.strip() for line in f.readlines()]

# function that takes a list of string and a slope as input and returns the number of trees encountered
def count_trees(lines, slope):
    x, y = 0, 0
    trees = 0
    while y < len(lines):
        if lines[y][x] == "#":
            trees += 1
        x = (x + slope[0]) % len(lines[0])
        y += slope[1]
    return trees

# function that takes a list of string and a list of slopes as input and returns the product of the number of trees encountered
def count_trees_product(lines, slopes):
    product = 1
    for slope in slopes:
        product *= count_trees(lines, slope)
    return product

# main
lines = read_file()
print("Part 1:", count_trees(lines, slopes[1]))
print("Part 2:", count_trees_product(lines, slopes))
